### PR TITLE
8269417: Minor clarification on NonblockingQueue utility

### PR DESCRIPTION
**Pull request by caoman**
Hi,

Could you review this change mainly based on the comments in https://github.com/openjdk/jdk/pull/4379?
I also added an assertion to ensure and better understand why the Atomic::store() in append() is correct.

Stress tested with fastdebug build with:
$ make test TEST="gtest:NonblockingQueueTestBasics gtest:NonblockingQueueTest" GTEST="REPEAT=-1"

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8269417](https://bugs.openjdk.java.net/browse/JDK-8269417): Minor clarification on NonblockingQueue utility

### Reviewers
 * [Kim Barrett](https://openjdk.java.net/census#kbarrett) (@kimbarrett - **Reviewer**) ⚠️ Review applies to e7890729ee284fd69bc86689868251d9a9f2f59e
 * [Ivan Walulya](https://openjdk.java.net/census#iwalulya) (@walulyai - Committer)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/4600/head:pull/4600` \
`$ git checkout pull/4600`

Update a local copy of the PR: \
`$ git checkout pull/4600` \
`$ git pull https://git.openjdk.java.net/jdk pull/4600/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 4600`

View PR using the GUI difftool: \
`$ git pr show -t 4600`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/4600.diff">https://git.openjdk.java.net/jdk/pull/4600.diff</a>

</details>

### CODE REVIEW

File: src/hotspot/share/utilities/nonblockingQueue.inline.hpp [hunk 1]
```diff
@@ -105,12 +103,23 @@ void NonblockingQueue<T, next_ptr>::append(T& first, T& last) {
...
+  if (is_old_tail_null ||
       // Try to install first as old_tail's next.
       !is_end(Atomic::cmpxchg(next_ptr(*old_tail), end_marker(), &first))) {
     // Install first as the new head if either
     // (1) the list was empty, or
     // (2) a concurrent try_pop claimed old_tail, so it is no longer in the list.
+    // Note that multiple concurrent push/append operations cannot modify
+    // _head simultaneously, because the Atomic::xchg() above orders these
+    // push/append operations so they perform Atomic::cmpxchg() on different
+    // old_tail. Thus, at most one Atomic::cmpxchg() can fail.
```
Comment (kimbarrett) [inline src/hotspot/share/utilities/nonblockingQueue.inline.hpp:116]: s/Thus, ... fail./Thus, the cmpxchg can only fail because of a concurrent try_pop./

**Review by kimbarrett — APPROVED** [review]
Looks good.  Just one suggestion in commentary.

**Review by kimbarrett — APPROVED** [review]
Looks good.

**Review by walulyai — APPROVED** [review]
lgtm!

**Pull request closed without merging**